Alt text for case list icons

nospame opened the pull request:
## Product Description
Adds an textbox to enter alt text on the key value mapping when the values are icons. For case list icons/clickable icons, this value is used in web apps to set the alt text on the icon, and for clickable icons is updated when the icon changes.
<img width="583" alt="image" src="https://github.com/dimagi/commcare-hq/assets/100609770/86ac5f30-d1c9-4819-b5aa-da3ec96faf32">

## Technical Summary
[USH-3933](https://dimagi.atlassian.net/browse/USH-3933)
https://github.com/dimagi/formplayer/pull/1536
https://github.com/dimagi/commcare-core/pull/1399, https://github.com/dimagi/commcare-core/pull/1400
Adds `alt_text` as a `LabelProperty` (`{'lang': 'value'}`) to `MappingItem`. UI is added in case list config to save this property. The property is used to generate an `alt_text` node in the suite for case list icons and clickable icons that lives within the `field` element, using the same calculation and variable names as its corresponding `value`, e.g.:
```xml
<alt_text>
  <text>
    <xpath function="if(starred = '0', $k0, if(starred = '1', $k1, ''))">
      <variable name="k0">
        <locale id="m0.case_short.case_starred_1.alt_text.k0"/>
      </variable>
      <variable name="k1">
        <locale id="m0.case_short.case_starred_1.alt_text.k1"/>
      </variable>
    </xpath>
  </text>
</alt_text>
```

## Feature Flag
Clickable icons behind CASE_LIST_CLICKABLE_ICON.

## Safety Assurance

### Safety story
Alt text is only added to the suite for appropriate column types "enum-image" and "clickable-icon" and is an optional element in Formplayer. Tested locally with configs for clickable icons and regular case list icons, including alternate language alt text.

### Automated test coverage
Automated tests added for alt text in suite and app strings. Additionally updated tests where `alt_text` is now an expected node in the suite, as with clickable icons.

### QA Plan
[QA-6080](https://dimagi.atlassian.net/browse/QA-6080)
This will get QA to ensure alt text is correctly saved and ultimately used by web apps.

### Rollback instructions
This should be deployed with or after Formplayer changes, otherwise apps with clickable icons or case list icons will not load correctly.

This PR can be reverted independently, but if Formplayer changes need to be reverted, this PR should be as well or it will break apps with now-invalid suite files.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-3933]: https://dimagi.atlassian.net/browse/USH-3933?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[QA-6080]: https://dimagi.atlassian.net/browse/QA-6080?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
UI to add alt text for icons/clickable icons

nospame committed:
Model and suite generation for alt text for icons

nospame committed:
Use altText from Formplayer in web apps

nospame committed:
Test alt text app strings

nospame committed:
Test alt text added to suite

nospame committed:
Bootstrap 5 templates + diffs

nospame committed:
Update test with alt text node

nospame committed:
Lint javascript

nospame committed:
Activate clickable icons with 'space' per WebAIM guidelines

See 'Button' https://webaim.org/techniques/keyboard/#testing

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -297,7 +297,10 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
 
         iconKeyAction: function (e) {
-            if (e.keyCode === 13) {
+            if (e.keyCode === 13 || e.keyCode === 32) {
+                if (e.keyCode === 32) {
+                    e.preventDefault();
+                }
```
This commit is unrelated to alt text but is accessibility-related and explicitly called out in making a "button" keyboard accessible. Ideally it would have been part of https://github.com/dimagi/commcare-hq/pull/33914 but I just now read the guideline and it seemed small enough to include here.

MartinRiese commented on corehq/apps/app_manager/app_strings.py:
```diff
@@ -166,6 +166,26 @@ def _create_module_details_app_strings(module, langs):
+                        ),
+                        clean_trans(item.value, langs)
+                    )
+                    yield (
```
If you added this `yield` in the original `for item in column.enum` behind an if statement you would not have to duplicate the code for the first one. Or extract it into a function.

nospame replied:
Good call, thanks https://github.com/dimagi/commcare-hq/pull/34044/commits/918093e3e64e9e9e4c15c2a6461f10a24ff47006

MartinRiese commented on corehq/apps/hqwebapp/static/hqwebapp/js/ui_elements/bootstrap3/ui-element-key-val-mapping.js:
```diff
@@ -39,8 +39,8 @@ hqDefine('hqwebapp/js/ui_elements/bootstrap3/ui-element-key-val-mapping', functi
             var c = s.charCodeAt(0);
-            if (c === 32) return '-';
-            if (c >= 65 && c <= 90) return '_' + s.toLowerCase();
+            if (c === 32) { return '-'; }
```
Not sure if we have a style guide for that but I would add the line breaks too when adding `{}`

nospame replied:
The closest to a _specific_ JS style guide I found (outside our linter settings, which think this is fine) is that Vellum refers to the javascript style guide from Airbnb. For what it's worth, that guide infers (but doesn't explicitly say) that single line conditionals and functions are fine. https://github.com/airbnb/javascript?tab=readme-ov-file#whitespace--block-spacing

nospame committed:
Refactor: more DRY alt text app strings

nospame committed:
Require 2.54 for alt text node

nospame committed:
Remove node from test now unneeded with 2.54 version check

nospame committed:
Don't show alt text UI if unsupported

nospame committed:
Bootstrap5 template + diffs

MartinRiese committed:
Merge branch 'master' into em/icons-alt-text

nospame committed:
Hide icon_alt_text template by default

nospame committed:
Update B5 template/diffs

nospame committed:
Add altText to case detail

nospame committed:
Change html structure of clickable icon to button > img

Intended to improve accessibility over role='button' since it is more universally supported

nospame committed:
Fix targeting for $moduleIcon

nospame committed:
Wait for FP response before reset icon

Prevents a flash of old icon before the new one is loaded, particularly on slow connections

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -293,15 +293,13 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
             const endpointId = temp.substring(temp.lastIndexOf('/') + 1);
             const endpointArg = urlTemplate.substring(urlTemplate.indexOf('?') + 1, urlTemplate.lastIndexOf('='));
-            e.target.className += " disabled";
+            $(e.target).closest('button.clickable-icon').addClass('disabled');
```
👍
